lib/sles4sap/console_redirection.pm

```diff
@@ -0,0 +1,174 @@
+Detects if login prompt appears and types the password.
+
+=cut
+
+sub connect_target_to_serial {
+    my (%args) = @_;
+    my $redirect_ip = $args{'target_ip'} // get_required_var('REDIRECT_TARGET_IP');
+    my $ssh_user = $args{'ssh_user'} // get_required_var('REDIRECT_TARGET_USER');
+
+    croak 'Missing "ssh_user" argument' unless $ssh_user;
+    croak 'Missing "redirect_ip" argument' unless $redirect_ip;
+    croak "OpenQA variable WORKER_VM_ID undefined. Run 'redirection_init()' first" unless get_var('WORKER_VM_ID');
+    croak "IP address '$redirect_ip' is not valid." unless verify_ip($redirect_ip);
+    croak 'Global variable "$searialdev" undefined' unless $serialdev;
+    croak "Console is already redirected to:" . script_output('hostname') if check_serial_redirection();
+
+    enter_cmd "ssh $ssh_opt $ssh_user\@$redirect_ip 2>&1 | tee -a /dev/$serialdev";
+    handle_login_prompt($ssh_user);
+    record_info('Redirect ON', "Serial redirection established to: $redirect_ip");
+}
+
+=head2 disconnect_target_from_serial
+
+    disconnect_target_from_serial($worker_id);
+
+Disconnects target from serial console by typing 'exit' command until host IP matches IP addr of the worker VM.
+
+=cut
+
+sub disconnect_target_from_serial {
+    my ($worker_id) = @_;
+    my $worker_machine_id = $worker_id // get_required_var('WORKER_VM_ID');
+    set_serial_term_prompt();
+    my $serial_redirection_status = check_serial_redirection($worker_machine_id);
+    while ($serial_redirection_status ne 0) {
+        enter_cmd('exit');    # Enter command and wait for screen start changing
+        $testapi::distri->{serial_term_prompt} = '';    # reset console prompt
+        wait_serial(qr/Connection.*closed./, timeout => 10);    # Wait for connection to close
+        wait_serial(qr/# |> /, timeout => 10);    # Wait for console prompt to appear
+        set_serial_term_prompt();    # after logout user might change and prompt with it.
+        $serial_redirection_status = check_serial_redirection($worker_machine_id);
+    }
+    record_info('Redirect OFF', "Serial redirection closed. Console set to: " . script_output('hostname'));
+}
+
+
+=head2 verify_ip
+
+    verify_ip($ip_address);
+
+B<ip_address> IP address
+
+Validates IP address format. Returns IP address if format is valied, 0 if invalid.
+
+=cut
+
+sub verify_ip {
+    my ($ip_address) = @_;
+    my $result = grep(/^$RE{net}{IPv4}$/, $ip_address) ? $ip_address : 0;
+    return $result;
+}
+
+=head2 check_serial_redirection
+
+    check_serial_redirection([$source_machine_id]);
+
+B<$source_machine_id> /etc/machine-id of the original machine where the console is redirected from
+
+Compares current machine-id to the worker VM ID either defined by WORKER_VM_ID variable or positional argument.
+Machine ID is used instead of IP addr since cloud VM IP might not be visible from the inside (for example via 'ip a')
+
+=cut
+
+sub check_serial_redirection {
+    my ($source_machine_id) = @_;
```

Comment:
can we use %args here too? Can we use a coherent name across all of the functions? here it is `source_machine_id` (that I like more) but in `disconnect_target_from_serial` it is `worker_id`, but I think they refer to the same magic number.

Reply:
Fixed